### DISCUSSION TITLE
psycopg3 batch vs row-wise insert (pyodbc's fast_executemany equivalent)

### DISCUSSION THREAD

**Discussion by aleewen** [psycopg/psycopg#872] [Q&A]
Hi devs,

I was wondering how inserting works with executemany in psycopg3.
By default with pyodbc, it sends the data row-by-row to the driver and this is slow. As a result, pyodbc has an argument called 'fast_executemany' that sends the entire rowset at once.
Does psycopg3 already do this by default? If not, how would I perform an insert a (polars) dataframe into a Postgres table all at once, rather than row-by-row?

Thanks

**Comment by dvarrazzo** [comment]
The fastest way to insert many rows in a postgres table is by using the COPY protocol, which is wrapped by [the copy methods](https://www.psycopg.org/psycopg3/docs/basic/copy.html#copy) in Psycopg 3.
